drivers/modem/gsm_ppp.c

```diff
@@ -430,6 +461,29 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
 	/* Attached, clear retry counter */
 	gsm->attach_retries = 0;
 
+	LOG_DBG("modem attach returned %d, %s", ret, "read RSSI");
+
+	/* Read connection quality (RSSI) */
+	LOG_INF("Read RSSI");
+	for (int i = 0; i < GSM_RSSI_RETRIES; i++) {
+		ret = modem_cmd_send_nolock(&gsm->context.iface,
+									&gsm->context.cmd_handler,
+									&read_rssi_cmd, 1,
+									"AT+CSQ",
+									&gsm->sem_response,
+									GSM_CMD_SETUP_TIMEOUT);
+		if (ret < 0) {
+			LOG_DBG("Not answer to RSSI readout, %s", "ignoring...");
+		}
+
+		if (gsm->context.data_rssi
+			&& gsm->context.data_rssi != GSM_RSSI_INVALID
+			&& gsm->context.data_rssi < GSM_RSSI_MAXVAL) {
+			break;
+		}
+		k_sleep(K_SECONDS(1));
```

Comment:
You need to re-design this loop because this function is called from system workqueue and it cannot sleep or wait something for a long time. That is why for example the `mux_setup()` does this in small pieces in a state machine.

Reply:
Thanks, I was not aware of this. I now built the logic into the existing state machine. Would this be a good solution?